### ISSUE TITLE
Something very wrong with float16 CPU implementation

### ISSUE THREAD

**Issue by ddpasa** [pytorch/pytorch#146508]
### 🐛 Describe the bug

There is something very wrong with the float16 implementation in the CPU backend.  It's much slower than the float32 implementation.  I created a very simple matmul test to reproduce the issue, posting results below:

```
PyTorch Version: 2.6.0+cpu
torch.backends.cpu.get_cpu_capability(): AVX512

500x500 matrix multiplication:
fp16: 0.10194063186645508 seconds
fp32: 0.001997709274291992 seconds
fp16 takes x51.0 times


1000x1000 matrix multiplication:
fp16: 0.9695210456848145 seconds
fp32: 0.007059574127197266 seconds
fp16 takes x137.3 times


2000x2000 matrix multiplication:
fp16: 7.783996105194092 seconds
fp32: 0.057384490966796875 seconds
fp16 takes x135.6 times
```

```
PyTorch Version: 2.7.0.dev20250205+cpu
torch.backends.cpu.get_cpu_capability(): AVX512

500x500 matrix multiplication:
fp16: 0.09653520584106445 seconds
fp32: 0.002435922622680664 seconds
fp16 takes x39.6 times


1000x1000 matrix multiplication:
fp16: 0.9120731353759766 seconds
fp32: 0.011063098907470703 seconds
fp16 takes x82.4 times


2000x2000 matrix multiplication:
fp16: 6.930181264877319 seconds
fp32: 0.07475137710571289 seconds
fp16 takes x92.7 times
```

System information: Intel Core i7-1065G7 CPU running on Linux 6.12.11

I am a regular user of llama.cpp, where I use all quants from fp16 to q4, and never observed extreme performance issues of this kind.

The benchmark script is below:

```
import time
import torch

print(f"PyTorch Version: {torch.__version__}")
print("torch.backends.cpu.get_cpu_capability():", torch.backends.cpu.get_cpu_capability())

def speedtest(N, precision):
    x = torch.randn(N, N, dtype=precision)
    y = torch.randn(N, N, dtype=precision)
    start = time.time()
    matmul_result = torch.matmul(x, y)
    end = time.time()
    return end - start

print('')
for N in (500, 1000, 2000):
    fp32 = speedtest(N, torch.float32)
    fp16 = speedtest(N, torch.float16)
    print(f'{N}x{N} matrix multiplication:\nfp16: {fp16} seconds\nfp32: {fp32} seconds\nfp16 takes x{round(fp16/fp32, 1)} times\n\n')
```

### Versions

Collecting environment information...
PyTorch version: 2.7.0.dev20250205+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Manjaro Linux (x86_64)
GCC version: (GCC) 14.2.1 20240910
Clang version: 19.1.7
CMake version: version 3.31.5
Libc version: glibc-2.40

Python version: 3.11.1 (main, Dec 21 2022, 10:10:41) [GCC 12.2.0] (64-bit runtime)
Python platform: Linux-6.12.11-1-MANJARO-x86_64-with-glibc2.40
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        39 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               8
On-line CPU(s) list:                  0-7
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Core(TM) i7-1065G7 CPU @ 1.30GHz
CPU family:                           6
Model:                                126
Thread(s) per core:                   2
Core(s) per socket:                   4
Socket(s):                            1
Stepping:                             5
CPU(s) scaling MHz:                   20%
CPU max MHz:                          3900,0000
CPU min MHz:                          400,0000
BogoMIPS:                             2996,00
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust sgx bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves split_lock_detect dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid sgx_lc fsrm md_clear flush_l1d arch_capabilities
Virtualization:                       VT-x
L1d cache:                            192 KiB (4 instances)
L1i cache:                            128 KiB (4 instances)
L2 cache:                             2 MiB (4 instances)
L3 cache:                             8 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-7
Vulnerability Gather data sampling:   Mitigation; Microcode
Vulnerability Itlb multihit:          KVM: Mitigation: Split huge pages
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                  Mitigation; Microcode
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.3
[pip3] onnxruntime==1.19.0
[pip3] torch==2.7.0.dev20250205+cpu
[conda] Could not collect

**Comment by ddpasa** [comment]
tagging @mingfeima because I've seen commits from him on this topic.

**Comment by malfet** [comment]
Interesting that slow-down numbers are somewhat similar on my Mac as well (running Apple silicon) Also I wonder if there are some sort of first time invocation problem in place.

**Comment by sanchitintel** [comment]
@ddpasa, FP16 matmul is supposed to be slow with PyTorch on x86_64 CPUs that don't support AVX512_FP16 ISA (and AMX_FP16 ISA), since FP16 is converted to FP32 before matmul, and then the matmul output is converted back to FP16.

Your CPU doesn't support AVX512_FP16 ISA either. Had you benchmarked FP16 matmul performance with llama.cpp on the same system?

**Comment by sanchitintel** [comment]
@ddpasa, to clarify, is the concern here that you expected FP16 matmul to be slower than FP32 matmul on your system, but not as slow as the benchmarking data you shared? Thanks!

**Comment by ddpasa** [comment]
> [@ddpasa](https://github.com/ddpasa), to clarify, is the concern here that you expected FP16 matmul to be slower than FP32 matmul on your system, but not as slow as the benchmarking data you shared? Thanks!

My concern is the massive performance drop (around x100 times) that happens when a model is run at float16 as opposed to float32.

Minimally there should be a warning of some kind that alerts the user of this inefficiency, so they can run at higher precision.  No need to run 100 times slower if you can avoid it.

But I was not expecting this to be this slow.  CPU does support F16C for example.  I will benchmark llama.cpp more closely, but I don't think there is such a dramatic performance drop from quantization.  I'll try to understand what they are doing in this scenario.

I also tried bfloat16 and the gap is much smaller, around x5 instead of x100.  Also, pytorch 2.7 is performing slightly better than 2.6, so there is also something in 2.7 that helps a little.

**Comment by ddpasa** [comment]
<details>

<summary> Updated benchmark including bfloat16 </summary>

```
PyTorch Version: 2.7.0.dev20250205+cpu
torch.backends.cpu.get_cpu_capability(): AVX512

500x500 matrix multiplication:
fp32: 0.002615213394165039 seconds
fp16: 0.10068058967590332 seconds
bf16: 0.010941267013549805 seconds
fp16 takes x38.5 times
bf16 takes x4.2 times

1000x1000 matrix multiplication:
fp32: 0.011064529418945312 seconds
fp16: 0.9702479839324951 seconds
bf16: 0.038317203521728516 seconds
fp16 takes x87.7 times
bf16 takes x3.5 times

2000x2000 matrix multiplication:
fp32: 0.08379673957824707 seconds
fp16: 7.412748098373413 seconds
bf16: 0.2097625732421875 seconds
fp16 takes x88.5 times
bf16 takes x2.5 times
```

</details>



<details>

<summary> Updated benchmark script with bfloat16 </summary>

```
import time
import torch

print(f"PyTorch Version: {torch.__version__}")
print("torch.backends.cpu.get_cpu_capability():", torch.backends.cpu.get_cpu_capability())

def speedtest(N, precision):
    x = torch.randn(N, N, dtype=precision)
    y = torch.randn(N, N, dtype=precision)
    start = time.time()
    matmul_result = torch.matmul(x, y)
    end = time.time()
    return end - start

print('')
for N in (500, 1000, 2000):
    print(f'{N}x{N} matrix multiplication:')
    fp32 = speedtest(N, torch.float32)
    print(f'fp32: {fp32} seconds')
    fp16 = speedtest(N, torch.float16)
    print(f'fp16: {fp16} seconds')
    bf16 = speedtest(N, torch.bfloat16)
    print(f'bf16: {bf16} seconds')
    print(f'fp16 takes x{round(fp16/fp32, 1)} times')
    print(f'bf16 takes x{round(bf16/fp32, 1)} times\n')
```

</details>

**Comment by ddpasa** [comment]
> Anyway, feel free to let me know if you consider that f16 gemm is important feature even on avx2 or avx512.

Yes, I think so.  Minimally there should be a warning for people when the super slow fallback path is triggered. No need to pay the x100 runtime hit.

I think there are a alot of devices in the market in this situation.  If there is any low hanging fruit that can be improved, that would be amazing.  I'm also happy to contribute where possible.

**Comment by sanchitintel** [comment]
@mingfeima, I think the concern here is not just that FP16 GEMM is slower on CPUs without AVX512_FP16 support, but that it's almost 100x slower.

Is it possible that for for large tensors, simply converting whole FP16 tensors to FP32 (rather than converting block-wise), computing FP32 GEMM with MKL, and converting the output back to FP16 may not be as slow as the current implementation in PyTorch?

@ddpasa, could you please run this modified script on your system? For more stable results (and better performance), please preload Intel OpenMP (libiomp5.so) & [tcmalloc](https://github.com/gperftools/gperftools). Thanks!

<details>
import time
import torch

print(f"PyTorch Version: {torch.__version__}")
print("torch.backends.cpu.get_cpu_capability():", torch.backends.cpu.get_cpu_capability())

def warmup(N):
    # JIT compile FP32 MKL GEMM kernel
    x = torch.randn(N, N)
    y = torch.randn(N, N)
    start = time.time()
    # Only one iteration is sufficient, but might as well run more
    for _ in range(10):
        matmul_result = torch.matmul(x, y)    

def speedtest(N, dtype=torch.float, cvt_fp16_to_fp32=False):
    x = torch.randn(N, N, dtype=dtype)
    y = torch.randn(N, N, dtype=dtype)
    start = time.time()
    for _ in range(10):
        if cvt_fp16_to_fp32:
            # Convert FP16 tensors to FP32 before matmul
            x = x.to(torch.float)
            y = y.to(torch.float)  
        matmul_result = torch.matmul(x, y)
        if cvt_fp16_to_fp32:
            # Convert back to FP16
            matmul_result = matmul_result.to(torch.float16)
    end = time.time()
    return end - start


print('')
for N in [512, 1024, 2048, 4096]:
    warmup(N)
    fp16 = speedtest(N, dtype=torch.float16, cvt_fp16_to_fp32=True)
    fp32 = speedtest(N, dtype=torch.float32)
    print(f'{N}x{N} matrix multiplication:\nfp16: {fp16} seconds\nfp32: {fp32} seconds\nfp16 takes x{round(fp16/fp32, 1)} times\n\n')

</details>

**Comment by ddpasa** [comment]
Hello @sanchitintel , this modified script makes a massive difference

```
PyTorch Version: 2.7.0.dev20250205+cpu
torch.backends.cpu.get_cpu_capability(): AVX512

512x512 matrix multiplication:
fp16: 0.009978055953979492 seconds
fp32: 0.007712125778198242 seconds
fp16 takes x1.3 times

1024x1024 matrix multiplication:
fp16: 0.12093734741210938 seconds
fp32: 0.0725409984588623 seconds
fp16 takes x1.7 times

2048x2048 matrix multiplication:
fp16: 0.5120499134063721 seconds
fp32: 0.5592305660247803 seconds
fp16 takes x0.9 times

4096x4096 matrix multiplication:
fp16: 4.40374493598938 seconds
fp32: 4.576500177383423 seconds
fp16 takes x1.0 times
```

**Comment by sanchitintel** [comment]
Thanks for confirming, @ddpasa! Some results can be attributed to run-to-run variation, but it looks like `converting FP16 tensors to FP32 -> MKL SGEMM -> converting FP32 output to FP16` is quite faster than the current implementation in PyTorch. FP16 <-> FP32 conversions are vectorized with AVX512F on your system (and F16C, in case of AVX2/AVX/SSE or when using ZMM registers is not necessary due to fewer elements requiring conversion).

@mingfeima, could we add such a workaround in ATen for systems with AVX512F or AVX2 (& F16C) ISA but not AVX512_FP16 or AMX_FP16 ISAs ( I'm not sure if we should also make such a change for older systems with AVX or SSE ISAs, although we could probably estimate the resulting performance of such systems by using the `MKL_ENABLE_INSTRUCTIONS` environment variable).

Thanks!

**Comment by ddpasa** [comment]
@mingfeima is there a fix here that I can test?  This is a huge performance drop and it would be great to have it patched.

**Comment by qflen** [comment]
Will PR a patch tomorrow with benchmarks.

**Comment by HasRahm** [comment]
Three additions from avx2-only hardware, since every report here so far is from AVX512 machines.

**1. `bfloat16` is affected too, not only `float16`.** Same matmul benchmark, Kaggle Xeon (avx2, no AVX512), torch 2.10.0+cpu:

| size | fp32 | bf16 | fp16 | bf16/fp32 | fp16/fp32 |
|---|---|---|---|---|---|
| 500² | 2.4 ms | 118.4 ms | 196.3 ms | 48.9× | 81.1× |
| 1000² | 27.2 ms | 1205.3 ms | 1753.9 ms | 44.4× | 64.5× |
| 2000² | 123.0 ms | 61407.8 ms | 61807.4 ms | **499.4×** | **502.7×** |

Reproduced on a second, unrelated avx2 machine (Intel Core 5 120U, torch 2.7.1+cpu): fp16 at 105× / 132× / **515×** for the same sizes.

**2. The penalty is far worse at 2000² than anything reported here** — 499× against the 39.6–137.3× in the original post. It is also non-monotonic in a way the AVX512 numbers are not: roughly flat 44–49× at 500² and 1000², then a ~10× jump at 2000². That looks like a cache-residency cliff in the convert → SGEMM → convert path rather than a constant conversion overhead.

**3. A measurement that supports @sanchitintel's diagnosis** (`FP16 → FP32 → MKL SGEMM → convert back`) and may explain why bf16 is hit at all. Conversion throughput alone, no matmul, 200M elements, 10 threads, on the Core 5 120U:

```
bfloat16 -> float32:   0.60 s   (0.67 GB/s of source)
float16  -> float32:   0.09 s   (4.49 GB/s of source)   6.7x faster
```

fp16→fp32 has an F16C hardware path, available on essentially every x86 CPU since Ivy Bridge. bf16→fp32 has no equivalent without `AVX512_BF16`, so it falls back to software. If the fix is in the convert step, bf16 needs it more than fp16 does on this hardware tier — which is the opposite of what the current issue title suggests.

**Why this matters in practice:** pre-quantized bitsandbytes checkpoints ship `bnb_4bit_compute_dtype: bfloat16`, and until huggingface/transformers#47661 lands that value cannot be overridden at load. On a CPU without `AVX512_BF16` every matmul then takes this path — a 14B LoRA step that runs ~116 s in fp32 does not complete at all. The visible symptom is "my machine is too slow", with nothing pointing at the dtype.

Environment: Intel Xeon @ 2.20 GHz (4 cores) and Intel Core 5 120U (12 threads), both avx2 only, no `avx512f` / `avx512_bf16` / `amx_bf16`; torch 2.10.0+cpu and 2.7.1+cpu.
